h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,172 @@
+import water.MemoryManager;
+import water.api.Handler;
+import water.util.ArrayUtils;
+import water.util.CollectionUtils;
+
+import java.util.*;
+
+/**
+ * Handling requests for various model trees
+ */
+public class TreeHandler extends Handler {
+    private static final int NO_CHILD = -1;
+
+    public TreeV3 getTree(final int version, final TreeV3 args) {
+        validateArgs(args);
+        final SharedTreeModel model = (SharedTreeModel) args.model.key().get();
+        if (model == null) throw new IllegalArgumentException("Unknown tree key: " + args.model.toString());
+
+        final SharedTreeModel.SharedTreeOutput sharedTreeOutput = (SharedTreeModel.SharedTreeOutput) model._output;
+        final CompressedTree auxCompressedTree = sharedTreeOutput._treeKeysAux[args.tree_number][args.tree_class].get();
+        final SharedTreeSubgraph sharedTreeSubgraph = sharedTreeOutput._treeKeys[args.tree_number][args.tree_class].get()
+                .toSharedTreeSubgraph(auxCompressedTree, sharedTreeOutput._names, sharedTreeOutput._domains);
+
+        final CompressedTreeFormat compressedTreeFormat = convertSharedTreeSubgraph(sharedTreeSubgraph);
+
+        args.left_children = compressedTreeFormat.leftChildren;
+        args.right_children = compressedTreeFormat.rightChildren;
+        args.root_node_number = compressedTreeFormat.rootNodeNumber;
+        return args;
+    }
+
+
+    /**
+     * @param args An instance of {@link TreeV3} input arguments to validate
+     */
+    private void validateArgs(TreeV3 args) {
+        if (args.tree_number < 0) throw new IllegalArgumentException("Tree number must be greater than 0.");
+        if (args.tree_class < 0) throw new IllegalArgumentException("Tree class must be greater than 0.");
+    }
+
+
+    /**
+     * Convers H2O-3's internal representation of a boosted tree in a form of {@link SharedTreeSubgraph} to a compressed format.
+     *
+     * @param sharedTreeSubgraph An instance of {@link SharedTreeSubgraph} to convert
+     * @return An instance of {@link CompressedTreeFormat} with some attributes possibly empty if suitable. Never null.
+     */
+    static CompressedTreeFormat convertSharedTreeSubgraph(final SharedTreeSubgraph sharedTreeSubgraph) {
+        Objects.requireNonNull(sharedTreeSubgraph);
+
+        final CompressedTreeFormat stf = new CompressedTreeFormat();
+        stf.rootNodeNumber = sharedTreeSubgraph.rootNode.getNodeNumber();
+
+        stf.leftChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        stf.rightChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        stf.thresholds = MemoryManager.malloc4f(sharedTreeSubgraph.nodesArray.size());
+
+        // Create a temporary nodeMap for faster search. Our nodes are not always numbered sequentially,
+        // thus searching in the array of nodes by index is not possible.
+        final Map<Integer, SharedTreeNode> nodeMap = createTreeNodeMap(sharedTreeSubgraph.nodesArray);
```

Comment:
this is not used for anything

Reply:
The whole createTreeNodeMap() should be deleted indeed.

Reply:
Done.